contracts/prover/MetaProver.sol

```diff
@@ -0,0 +1,212 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.26;
+
+import {IMetalayerRecipient, ReadOperation} from "@metalayer/contracts/src/interfaces/IMetalayerRecipient.sol";
+import {FinalityState} from "@metalayer/contracts/src/lib/MetalayerMessage.sol";
+import {TypeCasts} from "@hyperlane-xyz/core/contracts/libs/TypeCasts.sol";
+import {MessageBridgeProver} from "./MessageBridgeProver.sol";
+import {Semver} from "../libs/Semver.sol";
+import {IMetalayerRouter} from "@metalayer/contracts/src/interfaces/IMetalayerRouter.sol";
+
+/**
+ * @title MetaProver
+ * @notice Prover implementation using Caldera Metalayer's cross-chain messaging system
+ * @dev Processes proof messages from Metalayer router and records proven intents
+ */
+contract MetaProver is IMetalayerRecipient, MessageBridgeProver, Semver {
+    using TypeCasts for bytes32;
+    using TypeCasts for address;
+
+    /**
+     * @notice Constant indicating this contract uses Metalayer for proving
+     */
+    string public constant PROOF_TYPE = "Metalayer";
+
+    /**
+     * @notice Emitted when attempting to prove an already-proven intent
+     * @dev Event instead of error to allow batch processing to continue
+     * @param _intentHash Hash of the already proven intent
+     */
+    event IntentAlreadyProven(bytes32 _intentHash);
+
+    /**
+     * @notice Emitted when a batch of fulfilled intents is sent to the Metalayer router to be relayed to the source chain
+     * @param _hashes the intent hashes sent in the batch
+     * @param _sourceChainID ID of the source chain
+     */
+    event BatchSent(bytes32[] indexed _hashes, uint256 indexed _sourceChainID);
+
```

Comment:
Error handling patterns in MetaProver and HyperProver could be standardized to avoid duplication. Consider using a common library or inherited implementation.